File: Algorithms/Distribute Candies/distribute-candies.js
// Source : https://leetcode.com/problems/distribute-candies/description/
// Author : Fish.Yu
// Date   : 2017-12-25

/**
 * @param {number[]} candies
 * @return {number}
 */
var distributeCandies = function(candies) {
  let p = new Map()

  for (let item of candies) {
    p.set(item, ~~p.get(item) + 1)
  }

  let res = [...p.values()].sort((a, b) => a - b)
  
  return Math.min(res.length, candies.length >> 1)
};

File: Algorithms/Distribute Candies/distribute-candies.py
# Source : https://leetcode.com/problems/distribute-candies/description/
# Author : Fish.Yu
# Date   : 2017-12-26

class Solution:
    def distributeCandies(self, candies):
        """
        :type candies: List[int]
        :rtype: int
        """
        d = {}
        for item in candies:
            if item in d:
                d[item] += 1
            else:
                d[item] = 1

        res = sorted(list(d.values()))

        return min(len(res), len(candies) >> 1)